<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px" />

# Numpy Ladder Challenge - Notebook 2

_Author:_ Tim Book

# Climb the Ladder!
Our class moves quickly! Sometimes, it feels like we make leaps in logic that are a bit too big. In this ladder challenge, we'll learn some core math concepts, some linear algebra, and the `numpy` library. Problems in this notebook start out easy and progressively get harder, so that the next rung of the Python ladder is always within reach.

Additionally, since not all of the topics discussed in this ladder challenge are explicitly taught in our course, these problems come with many more hints, tips, suggestions, and even sometimes a mini-lesson. You are encouraged to Google frequently throughout the lesson. In many ways, this ladder is meant to be a challenge as well as educational in its own right.

**Remember our one rule: NO LOOPS! None of the exercises in this notebook require a loop. If you use a loop to solve any of these problems, you are solving the problem incorrectly.**

0) Import numpy in the usual way

In [1]:
import numpy as np

# Section III - Simulation
In the next section, we'll use functions within the `np.random` submodule. You can find documentation [here](https://docs.scipy.org/doc/numpy-1.14.0/reference/routines.random.html).

27) Generate 10,000 random numbers between 0 and 1 and assign them to a variable. To verify you've simulated the data properly, make sure the mean is approximately 0.5.

In [4]:
np.random.seed(3)
random_10000 = np.random.sample(10_000)
np.mean(random_10000)

0.49678509429483986

28) What proportion of these numbers is below 0.1?

In [10]:
sum(random_10000 < 0.1)/10_000

0.0976

29) What proportion of these numbers is above 0.8?

In [11]:
sum(random_10000 > 0.8)/10_000

0.1973

30) What proportion of these numbers is above 0.2 but below 0.3?

In [12]:
sum((random_10000 > 0.2) & (random_10000 < 0.3))/10_000

0.1065

31) Generate 10,000 random numbers between 2 and 4. To verify you've simulated the data properly, find the mean and make sure it is approximately what you expect.

* _Hint:_ There is no `numpy` function for this specifically. How can you use the function you just used to generate this?

> The expected value is 3

In [37]:
# looks like there is a function called unform could do this
np.random.seed(3)
between_2_4_a = np.random.uniform(2,4,10_000)
np.mean(between_2_4_a)

2.9935701885896795

In [38]:
# however, i used a second method does the same thing
np.random.seed(3)
between_2_4_b = np.random.choice([2,3],10_000) + np.random.sample(10_000)
np.mean(between_2_4_b)

2.9916334753460743

32) What proportion of these numbers is between 2.4 and 2.6?

In [41]:
sum((between_2_4_a > 2.4) & (between_2_4_a < 2.6)) / 10_000

0.1065

33) Generate 100,000 random standard normal (i.e., mean 0 standard deviation 1) numbers. Again, find the mean to verify you've done this properly.

In [48]:
np.random.seed(9)
random_nor_10000 = np.random.randn(10_000)

np.mean(random_nor_10000)

-0.008540271041175544

34) What proportion of these numbers is negative?

In [49]:
sum(random_nor_10000 < 0) / 10_000

0.5083

35a) What proportion of these numbers is between -1 and 1?

In [50]:
sum((random_nor_10000 > -1) & (random_nor_10000 < 1)) / 10_000

0.6885

35b) What proportion of these numbers is between -2 and 2?

In [51]:
sum((random_nor_10000 > -2) & (random_nor_10000 < 2)) / 10_000

0.9536

35c) What proportion of these numbers is between -3 and 3? Have you seen your last 3 solutions before? (If you've taken an intro stats course in college before, you will have.)

In [52]:
sum((random_nor_10000 > -3) & (random_nor_10000 < 3)) / 10_000

0.9967

## Answer:
> Yes, I have seen this before. This 3 proportions exactly follows the empirical rule, 68-95-99.7 

### For the next few problems, we will be playing Rock-Paper-Scissors
If you are unfamiliar with the game Rock-Paper-Scissors, it features two combatants choosing one of three hand motions: rock, paper, or scissors. Rock beats scissors, scissors beats paper, and paper beats rock. Two friends are playing: Karen and Tom. Unbeknownst to them, you've been studying and recording both of their play patterns. Karen chooses rock 40% of the time, paper 10% of the time, and scissors 50% of the time. For Tom, it's rock 10%, paper 60%, scissors 30%. Who wins the most often?

**This is an extremely difficult question.** We will get to the answer in a few guided steps. You'll want to use `np.random.choice()` to help you through this.

36) Create vectors `p_karen` and `p_tom` that represent their respective probabilities for rock, paper, and scissors.

In [53]:
p_karen = np.array([0.4,0.1,0.5])
p_tom = np.array([0.1,0.6,0.3])

37) Simulate 5 games. Who wins the majority of them? Just eyeball this one. (No one wins a draw.)

I guess Karen will win the majority of them. beacuse Karen's scissors probabiliy is the higherst, and Tom has the highest probability for paper, scissors beats paper.

38) Let's write a function to handle _one game at a time_. Write a function called `rps` that takes two arguments: `karen` and `tom` that will be either `"R"`, `"P"`, or `"S"`. The function will return `"K"`, `"T"`, or `"D"`, representing Karen, Tom, or a draw. That is, the function should give the following results:

* `rps("R", "P") ==> "T"`
* `rps("R", "S") ==> "K"`
* `rps("R", "R") ==> "D"`


* _Hint:_ Your answer will be a mess of `if`/`elif` statements.

In [83]:
def rps(karen,tom):
    player = {karen:'K', tom:'T'}
    comb = list(player.keys())
    
    if ('P' in comb) and ('R' in comb):
        return player['P']
    elif ('S' in comb) and ('P' in comb):
        return player['S']
    elif ('S' in comb) and ('R' in comb):
        return player['R']
    else:
        return 'D'

In [91]:
rps("R", "P")

'T'

In [92]:
rps("R", "S")

'K'

In [94]:
rps("S","P")

'K'

In [93]:
rps("R", "R")

'D'

39) As it stands now, the function you have written cannot handle vector data. Luckily, `numpy` gives us a function that allows us to vectorize any function we want. Use `np.vectorize` to create `rps_vectorized`, the vectorized version of `rps`. Skim the docs [here](https://docs.scipy.org/doc/numpy/reference/generated/numpy.vectorize.html).

In [98]:
rps_vectorized = np.vectorize(rps)

40) Simulate 1,000,000 (yes, one million) games. How often does Karen win? You can find the results by:

1. Replicating your solution to problems 37 and 38, except for one million instead of 5.
1. Using the function you made in problem 39.


* _Note 1:_ These probabilities are relatively difficult to figure out by hand. Some probabilities are best discovered by simulation. Another way of asking the above question is "What is the probability of Karen winning?")
* _Note 2:_ Look what we did! We used vectorization to completely eliminate the need for a loop! You _could_ have solved this problem with a loop, but it would have taken significantly more computer time.

In [110]:
np.random.seed(3)

k_random = np.random.choice(['R','P','S'],1_000_000,p=p_karen)
t_random = np.random.choice(['R','P','S'],1_000_000,p=p_tom)

result = rps_vectorized(k_random,t_random)

sum(result=='K')/1_000_000

0.429773

> Karen wins 42.98% of the time.

### Regression Simulation
Next, suppose we're trying to simulate some fake data for a regression problem we wish to give to our students. We wish to simulate data for the following equation:

$$y = 1000 + 200x + \varepsilon$$

Where $\varepsilon \sim N(0, 20)$ (that is, $\varepsilon$ is normally distributed with mean 0 and standard deviation 20).

41) Generate 10 $x$s from the $N(50, 10)$ distribution. (That is, the normal distribution with mean 50 and standard deviation 10).

In [113]:
np.random.seed(3)
xs = np.random.normal(50,10,10)

42) Generate 10 $\varepsilon$s from the appropriate distribution described above.

In [114]:
np.random.seed(3)
es = np.random.normal(0,20,10)

43) Simulate the $y$s as described above using the two vectors you made in the previous two problems.

In [116]:
y = 1000 + 200*xs + es

In [117]:
y

array([14613.02951633, 11881.74989803, 11194.92488551,  7235.7447392 ,
       10439.67583092, 10283.38686188, 10832.86220741,  9733.45863282,
       10911.48729867, 10036.01957867])

# Section IV: Matrices
_Tiny note:_ In general, we have told you it's against Python's style guide to use capital letters when defining variables. The one exception we make is when our variables represent mathematical objects. So feel free to name the following variables things like `A`, `B`, etc.!

44) Create the following matrix:

$$
A =
\begin{bmatrix}
3 & -1 & 5 \\
-2 & 0 & 8 \\
4 & 5 & -7
\end{bmatrix}
$$

In [119]:
A = np.array([[3,-1,5],
              [-2,0,8],
              [4,5,-7]])

In [120]:
A

array([[ 3, -1,  5],
       [-2,  0,  8],
       [ 4,  5, -7]])

45) Create the following matrix:

$$
B =
\begin{bmatrix}
-3 & 8 & 2 \\
2 & -3 & 5 \\
0 & 6 & -2
\end{bmatrix}
$$

In [121]:
B = np.array([[-3,8,2],
              [2,-3,5],
              [0,6,-2]])

In [122]:
B

array([[-3,  8,  2],
       [ 2, -3,  5],
       [ 0,  6, -2]])

46) Use `np.eye()` to create the following matrix:

$$
I = I_3 =
\begin{bmatrix}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 0 & 1 \\
\end{bmatrix}
$$

In [123]:
I = np.eye(3)

In [126]:
I

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

47) Triple every element of `B`. (Do not reassign.)

In [127]:
B * 3

array([[-9, 24,  6],
       [ 6, -9, 15],
       [ 0, 18, -6]])

48) Index `A` in order to get me -7.

In [130]:
A[2][2]

-7

49) Index `A` in order to get me 0.

In [131]:
A[1][1]

0

50) Index `A` to get me $\begin{bmatrix} -2 & 0 & 8 \end{bmatrix}$

In [144]:
A[1]

array([-2,  0,  8])

51) Index `A` to get me
$$
\begin{bmatrix}
-1 \\ 0 \\ 5
\end{bmatrix}
$$

(You can ignore the fact that it may print out as a row vector)

In [145]:
A[:,1]

array([-1,  0,  5])

52) Index `A` to get me
$$
\begin{bmatrix}
0 & 8 \\
5 & -7 
\end{bmatrix}
$$

In [147]:
A[1:,1:]

array([[ 0,  8],
       [ 5, -7]])

53) Redefine the middle column of `A` to be
$$
\begin{bmatrix}
-2 \\ 1 \\ 7
\end{bmatrix}
$$

In [149]:
A[:,1] = [-2,1,7]

In [150]:
A

array([[ 3, -2,  5],
       [-2,  1,  8],
       [ 4,  7, -7]])

54) Index `B` to define the following matrix:
$$
C = 
\begin{bmatrix}
8 & 2 \\
-3 & 5 \\
6 & -2
\end{bmatrix}
$$

In [176]:
C = B[:,1:]

In [177]:
C

array([[ 8,  2],
       [-3,  5],
       [ 6, -2]])

55) What is $A + B$?

In [153]:
A + B

array([[ 0,  6,  7],
       [ 0, -2, 13],
       [ 4, 13, -9]])

56) What is $2A - 3B$?

In [155]:
2*A - 3*B

array([[ 15, -28,   4],
       [-10,  11,   1],
       [  8,  -4,  -8]])

57) What is the elementwise product of $A$ and $B$?

In [161]:
A*B

array([[ -9, -16,  10],
       [ -4,  -3,  40],
       [  0,  42,  14]])

58) What is $AB$?

In [165]:
np.matmul(A,B)

array([[-13,  60, -14],
       [  8,  29, -15],
       [  2, -31,  57]])

59) What is $BA$? And why isn't it the same as $AB$?

In [166]:
np.matmul(B,A)

array([[-17,  28,  35],
       [ 32,  28, -49],
       [-20,  -8,  62]])

**Answer**:
> The order in matrix multiplication matters.

60) What is $AI$, and is it equal to $IA$? Does this product look familiar?

In [173]:
np.dot(A,I)

array([[ 3., -2.,  5.],
       [-2.,  1.,  8.],
       [ 4.,  7., -7.]])

In [174]:
np.dot(I,A)

array([[ 3., -2.,  5.],
       [-2.,  1.,  8.],
       [ 4.,  7., -7.]])

**Answer**:
> Yes, they are same. I is identity matrix, so IA = AI = A
    

61) What is $AC$?

In [178]:
A.dot(C)

array([[ 60, -14],
       [ 29, -15],
       [-31,  57]])

62) Why do we get an error when calculating $CA$?

**Answer**:
> Because matrix multiplcation require first matrix's column number equals second matrix's row number.

63) What is $A^TA$? Note that you answer will be a _diagonal matrix_, that is, a matrix that is equal to its transpose.

In [186]:
np.transpose(A).dot(A)

array([[ 29,  20, -29],
       [ 20,  54, -51],
       [-29, -51, 138]])

64) What is $A^TB$?

In [188]:
A.transpose().dot(B)

array([[-13,  54, -12],
       [  8,  23, -13],
       [  1, -26,  64]])

65) What is $A^{-1}$ (the inverse of $A$).

In [189]:
np.linalg.inv(A)

array([[ 0.2       , -0.06666667,  0.06666667],
       [-0.05714286,  0.13015873,  0.10793651],
       [ 0.05714286,  0.09206349,  0.0031746 ]])

66) What is $AA^{-1}$? Does it look familiar?

* _Hint:_ Maybe call an `np.round()` on your result.

In [190]:
np.round(A.dot(np.linalg.inv(A)))

array([[ 1.,  0., -0.],
       [-0.,  1.,  0.],
       [ 0., -0.,  1.]])

67) What is $(B + A^TA)^{-1}A^TC$?

In [191]:
np.linalg.inv(B + A.transpose().dot(A)).dot(A.transpose().dot(C))

array([[ 3.00215282, -0.28385833],
       [-0.62588766,  0.33839645],
       [ 0.24189329,  0.52202903]])